### PR TITLE
[FRONTEND][TENSORFLOW] GPU support for tensorflow models.

### PR DESCRIPTION
**Pull request by srkreddy1238**
By layout conversion of incompatible operations (conv, pool  ..etc) to NCHW.

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

**Comment by srkreddy1238** [comment]
As expected significant improvement by supporting GPU for tensorflow models 

InceptionV1 end to end average inference show below results.

CPU: Xeon @2.20GHz - 2 Core :: 3.01 sec 
GPU: NVIDIA Tesla K80  - 1 Core :: 0.0329 sec

**Comment by masahi** [comment]
@srkreddy1238 are you adding layout trasnform ops into the network? I've also done NCHW -> NHWC conversion for a whole network in the past. My approach was as follows: 

* First, load a standard NCHW layout model from frontend into a nnvm symbol
* Overwrite "layout" attributes and other layout dependent attribute using [this method](https://github.com/dmlc/tvm/blob/master/nnvm/python/nnvm/_ctypes/symbol.py#L94)
* Transpose filter weights for NHWC layout

This way, no in-network layout conversion is necessary and I can operate on NHWC input throughout.

**Comment by masahi** [comment]
You can also overwrite "axis" and "shape" attributes. I don't about squeeze op, but I can convert reshape op 
 between NCHW and NHWC.

**Comment by srkreddy1238** [comment]
@masahi @yzhliu @FrozenGene welcome to review.

@masahi altering axis, shape for layout independent operations is not a generalized solution and it's a never-ending list in this way. Hence I propose transpose those few operations instead.

What do you think?

**Comment by masahi** [comment]
yeah I think your approach is a good compromise. If users want their tf model to be high performant, their model should be in NCHW layout in the first place.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 1]
```diff
@@ -158,41 +164,61 @@ def _impl(inputs, attr, params):
...
+        input_shape = attr['_input_shapes'][inputs[0]][0]
+        weights_shape = params[inputs[1].list_output_names()[0]].shape
+
+        if attr['_target_layout'] == "NCHW" and attr['data_format'] == "NHWC":
+            input_shape = [input_shape[ii] for ii in (0, 3, 1, 2)]
+            inputs[0] = _sym.transpose(inputs[0], axes=(0, 3, 1, 2))
+            if opname == 'conv':
+                weights_shape = [weights_shape[ii] for ii in (3, 2, 0, 1)]
+                inputs[1] = _sym.transpose(inputs[1], axes=(3, 2, 0, 1))
+            else:
```
Comment (masahi) [inline nnvm/python/nnvm/frontend/tensorflow.py]: what is this else branch for?

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 2]
```diff
@@ -121,19 +122,24 @@ def _pooling(name):
...
-            attr['kernel_shape'] = (attr['ksize'][1], attr['ksize'][2])
-        elif attr['data_format'] == 'NCHW':
-            attr['kernel_shape'] = (attr['ksize'][2], attr['ksize'][3])
-        else:
-            raise TypeError("Unsupported data_format type : {}".format(attr['data_format']))
+        input_shape = attr['_input_shapes'][inputs[0]][0]
+
+        if attr['_target_layout'] == "NCHW" and attr['data_format'] == "NHWC":
+            tmp_shape = attr['_input_shapes'][inputs[0]][0]
+            tmp_shape = [tmp_shape[ii] for ii in (0, 3, 1, 2)]
```
Comment (masahi) [inline nnvm/python/nnvm/frontend/tensorflow.py]: input_shape = [tmp_shape[ii] for ii in (0, 3, 1, 2)]

**Comment by srkreddy1238** [comment]
Depthwise convolution.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 3]
```diff
@@ -121,19 +122,24 @@ def _pooling(name):
...
+            input_shape = tmp_shape
+            inputs[0] = _sym.transpose(inputs[0], axes=(0, 3, 1, 2))
+            attr['data_format'] = "NCHW"
+            flip_layout = True
+
+        attr['kernel_shape'] = (attr['ksize'][1], attr['ksize'][2])
```
Comment (masahi) [inline nnvm/python/nnvm/frontend/tensorflow.py]: does it work for NCHW layout?

Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Done

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 4]
```diff
@@ -129,11 +133,17 @@ def _impl(inputs, attr, params):
...
 
+        if attr['_target_layout'] == "NCHW" and attr['data_format'] == "NHWC":
+            tmp_shape = attr['_input_shapes'][inputs[0]][0]
+            input_shape = [tmp_shape[ii] for ii in (0, 3, 1, 2)]
+            inputs[0] = _sym.transpose(inputs[0], axes=(0, 3, 1, 2))
+            attr['data_format'] = "NCHW"
+            flip_layout = True
+
         # Fix strides
         attr['strides'] = (attr['strides'][1], attr['strides'][2])
```
Comment (masahi) [inline nnvm/python/nnvm/frontend/tensorflow.py]: what about this one? does this work for NCHW?

**Comment by srkreddy1238** [comment]
@masahi 

We never integrated end-to-end pure NCHW model support till now. 
I have resnet50_v2 which is pure NCHW from tensorflow working which is yet to PR.

This has few other fixes to frontend. Here I will try adding few test cases for NCHW exclusively.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 1 above)
Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: for depthwise comvolution support.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 2 above)
Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Done

**Comment by masahi** [comment]
@srkreddy1238 ok, I was about to ask NCHW conv test for tf frontend, but if this is coming then it's ok.

**Comment by masahi** [comment]
@srkreddy1238 I have one question unanswered. Otherwise looks good.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 4 above)
Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Strides also require handling for NCHW. This is handled in my other branch where I am working on NCHW with resnet models.

Comment (masahi) [reply nnvm/python/nnvm/frontend/tensorflow.py]: ok

**Comment by srkreddy1238** [comment]
@yzhliu , @tqchen can you merge this ?
I have following PR on top of this.

**Comment by tqchen** [comment]
Thanks @masahi @srkreddy1238 this is now merged

**Comment by FrozenGene** [comment]
> @srkreddy1238 are you adding layout trasnform ops into the network? I've also done NCHW -> NHWC conversion for a whole network in the past. My approach was as follows:
> 
> * First, load a standard NCHW layout model from frontend into a nnvm symbol
> * Overwrite "layout" attributes and other layout dependent attribute using [this method](https://github.com/dmlc/tvm/blob/master/nnvm/python/nnvm/_ctypes/symbol.py#L94)
> * Transpose filter weights for NHWC layout
> 
> This way, no in-network layout conversion is necessary and I can operate on NHWC input throughout. This is a kind of hack, but it worked for me.
> 
> In your case, you can overwrite all NHWC layout attributes, run NCHW network as usual, and then convert the output to NHWC if necessary.

Could you expain more detail? I want to know a little more how to do it use your mentioned method, for example Reshape op.

**Pull request merged**